plugins/imagebase/plugin.js

```diff
@@ -0,0 +1,149 @@
+		 * {@link CKEDITOR.plugins.widget.definition}.
+		 * It adds several parts of image and implements the {@link CKEDITOR.plugins.widget.definition#upcast} callback.
+		 * This callback should not be overwritten.
+		 *
+		 * @abstract
+		 * @class CKEDITOR.plugins.imagebase.imageWidgetDefinition
+		 * @mixins CKEDITOR.plugins.widget.definition
+		 */
+		return {
+			/**
+			 * @inheritdoc
+			 */
+			pathName: editor.lang.imagebase.pathName,
+
+			/**
+			 * @inheritdoc
+			 */
+			template: defaultTemplate,
+
+			/**
+			 * @inheritdoc
+			 */
+			allowedContent: {
+				img: {
+					attributes: '!src,alt,width,height'
+				},
+				figure: {
+					classes: '!' + joinClasses( options.basicClass, options.additionalClasses )
+				},
+				figcaption: true
+			},
+
+			/**
+			 * @inheritdoc
+			 */
+			requiredContent: 'figure(!' + options.basicClass + ')',
+
+			/**
+			 * @inheritdoc
+			 */
+			editables: {
+				caption: {
+					selector: 'figcaption',
+					pathName: editor.lang.imagebase.pathNameCaption,
+					allowedContent: options.captionAllowedContent
+				}
+			},
+
+			/**
+			 * @inheritdoc
+			 */
+			parts: {
+				image: 'img',
+				caption: 'figcaption'
+			},
+
+			/**
+			 * Image widget definition overwrites the {@link CKEDITOR.plugins.widget.definition#upcast} property.
+			 * This should not be changed.
+			 *
+			 * @property {Function}
+			 */
+			upcast: function( element ) {
+				// http://dev.ckeditor.com/ticket/11110 Don't initialize on pasted fake objects.
+				if ( element.attributes[ 'data-cke-realelement' ] ) {
+					return;
+				}
+
+				if ( element.name === 'figure' && element.hasClass( options.basicClass ) ) {
+					return element;
+				}
+			},
+
+			/**
+			 * Invokes user defined widget initialization method.
+			 *
+			 * @property {Function}
+			 */
+			init: function() {
+				if ( options.init ) {
+					options.init.call( this );
+				}
+			}
+		};
+	}
+
+	CKEDITOR.plugins.add( 'imagebase', {
+		requires: 'widget',
+		lang: 'en'
+	} );
+
+	/**
+	 * Namespace providing a set of helper functions for working with image widgets.
+	 *
+	 * @since 4.8.0
+	 * @singleton
+	 * @class CKEDITOR.plugins.imagebase
+	 */
+	CKEDITOR.plugins.imagebase = {
+		/**
+		 * Registers a new widget definition based on passed options.
+		 *
+		 * @since 4.8.0
+		 * @param {CKEDITOR.editor} editor Editor that will get the definition registered.
+		 * @param {String} name Widget name.
+		 * @param {Object} options Widget's options.
+		 * @param {String} options.basicClass Primary class for widget's element, that allows to properly
+		 * identify the widget.
+		 * @param {String[]} options.additionalClasses Additional classes that could be applied to widget's
+		 * element.
+		 * @param {CKEDITOR.filter.allowedContentRules} options.captionAllowedContent Allowed content inside
+		 * image's caption.
+		 * @param {Function} options.init Widget's initialization method.
+		 */
+		createWidget: function( editor, name, options ) {
```

Comment:
Let's stay consistent with `addFooWidget` [that we started in upload plugin](https://github.com/ckeditor/ckeditor-dev/blob/8f1d80f/plugins/uploadwidget/plugin.js#L535). So it should be `addImageWidget`.

Reply:
Current API is very, very rigid and little extensive. Instead I think it should allow you to customize any of the about-to-be-created widget definition.

In that case I can see `createWidgetDefinition` returning a plain/"ideal" definition for image base, and then within `addImageWidget` it would call `CKEDITOR.object.extend` if custom overriding rule was provided.

Reply:
`options` argument should simply allow to customize widget definition. So say it would be a  `definition` argument, an instance of `CKEDITOR.plugins.imagebase.imageWidgetDefinition`.